tests/unit/deferred.py

```diff
@@ -0,0 +1,222 @@
+
+            local_db.session.commit()
+            local_db.session.close()
+
+            if self.last_count is None:
+                self.last_count = current_count
+            else:
+                assert current_count == self.last_count - 1, \
+                  "network daemon did not commit previous change!"
+                self.last_count = current_count
+
+        def revert_port(self, port):
+            raise SwitchError('switch failed')
+
+    DeferredTestSwitch_.__name__ = 'DeferredTestSwitch'
+    DeferredTestSwitch = DeferredTestSwitch_
+
+
+@pytest.fixture()
+def switch(_deferred_test_switch_class):
+    return DeferredTestSwitch(
+        label='switch',
+        hostname='http://example.com',
+        username='admin',
+        password='admin',
+    ).session()
+
+
+def new_nic(name):
+    from hil.ext.obm.mock import MockObm
+    return model.Nic(
+        model.Node(
+            label='node-99',
+            obm=MockObm(
+                type="http://schema.massopencloud.org/haas/v0/obm/mock",
+                host="ipmihost",
+                user="root",
+                password="tapeworm")),
+        name,
+        '00:11:22:33:44:55')
+
+
+@pytest.fixture()
+def network():
+    project = model.Project('anvil-nextgen')
+    return model.Network(project, [project], True, '102', 'hammernet')
+
+pytestmark = pytest.mark.usefixtures('configure',
+                                     'fail_on_log_warnings')
+
+
+def test_apply_networking(switch, network, fresh_database):
+    '''Test to validate apply_networking commits actions incrementally
+
+    This test verifies that the apply_networking() function in hil/deferred.py
+    incrementally commits actions, which ensures that any error on an action
+    will not require a complete rerun of the prior actions (e.g. if an error
+    is thrown on the 3rd action, the 1st and 2nd action will have already been
+    committed)
+
+    The test also verifies that if a new networking action fails, then the
+    old networking actions in the queue were commited.
+    '''
+    nic = []
+    action = []
+    # initialize 3 nics and networking actions
+    for i in range(0, 3):
+        interface = 'gi1/0/%d' % (i)
+        nic.append(new_nic(str(i)))
+        nic[i].port = model.Port(label=interface, switch=switch)
+        action.append(model.NetworkingAction(nic=nic[i], new_network=network,
+                                             channel='vlan/native',
+                                             type='modify_port'))
+
+    # this makes the last action invalid for the test switch which will raise
+    # an exception.
+    action[2] = model.NetworkingAction(nic=nic[2],
+                                       new_network=None,
+                                       channel='',
+                                       type='revert_port')
+    db.session.add(action[0])
+    db.session.add(action[1])
+    db.session.add(action[2])
```

Comment:
This can be a for loop. It only saves one line, but makes the pattern more clear.